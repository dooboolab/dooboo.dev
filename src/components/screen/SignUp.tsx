import React, { Component } from 'react';
import styled from 'styled-components';

const Container = styled.div`
  flex: 1;
  background-color: transparent;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

interface IProps {
  history?: any;
}

function Page(props: IProps) {
  return (
    <Container>
      <div
        data-testid='myText'
      >Sign Up</div>
    </Container>
  );
}

export default Page;
