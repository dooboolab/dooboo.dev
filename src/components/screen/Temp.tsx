import React, { Component } from 'react';
import styled from 'styled-components';
import Button from '@material-ui/core/Button';

const Container = styled.div`
  display: flex;
  height: 100vh;
  justify-content: center;
  align-items: center;
  background: ${(props) => props.theme.background};
  padding: 50px;
`;

interface IProps {
  history?: any;
}

function Temp(props: IProps) {
  const { history } = props;
  return (
    <Container>
      <Button
        onClick={() => history.goBack()}
      >back to tab page</Button>
    </Container>
  );
}

export default Temp;
