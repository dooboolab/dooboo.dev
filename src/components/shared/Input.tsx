import React, { useEffect, memo, NamedExoticComponent } from 'react';
import { ErrorMessage } from 'formik';
import styled from 'styled-components';
import { FormLabel, FormError, FormInput } from '../ui/Forms';

export const FormContainer = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 4rem;
  align-self: flex-end;
`;

export const FormInnerContainer = styled.div`
  display: flex;
  flex-direction: row-reverse;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: center;
  margin: 1rem auto 0.4rem auto;
`

export const getLabelCapitalize = (label: string) => `${label.charAt(0).toUpperCase()}${label.slice(1)}`;

export const SimpleFormInput: NamedExoticComponent<any> = memo(({ field, form, type, width, receivedEmail, setEmail }) => {
  const { name, ...restFields } = field;
  const  { touched: touches, errors, setFieldValue } = form;
  const isTouched = touches && touches[name];
  const error = errors && errors[name]
  const rest = { type, width, isTouched, error, ...restFields }
  useEffect(() => {
    if (name === 'email' && setEmail && receivedEmail) {
      setFieldValue(name, receivedEmail);
    }
    return () => {
      if (setEmail) {
        setEmail('');
      }      
    }
  }, [setEmail, receivedEmail]);
  
  return (
    <FormContainer>
      <FormInnerContainer>        
        <FormInput id={name} {...rest} />
        <FormLabel data-testid={`label-${name}`} htmlFor={name}>{(name && name.length) && getLabelCapitalize(name)}</FormLabel>
      </FormInnerContainer>
      <ErrorMessage name={name}>
        {err => <FormError data-testid={`errors-${name}`}>{err}</FormError>}
      </ErrorMessage>
    </FormContainer>
  )
})