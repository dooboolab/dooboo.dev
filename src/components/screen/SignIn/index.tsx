import React, { useState, useContext, useCallback, Dispatch } from 'react';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import Swal from 'sweetalert2';
import withReactContent from 'sweetalert2-react-content';
import { History } from 'history';

import { AppContext } from '../../../contexts';
import {
  Wrapper,
  SignContainer,
  FormInnerWrapper,
  BottomTextBtnContainer,
  BottomTextButton,
} from '../SignUp/styles';
import { SimpleFormInput } from '../../shared/Input';
import Button from '../../shared/Button';

export interface IProps {
  history?: History<{ email: string }>
}

const ReactSwal = withReactContent(Swal);

export const signInList = [{
  label: 'email',
  type: 'text',
}, {
  label: 'password',
  type: 'password',
}];

const SwalForForgotPW = (pwReset: (email: string) => Promise<void>, setEmail: Dispatch<string>) => {
  ReactSwal.fire({
    title: 'Input your email address.',
    text: 'We\'ll send an email which has a link for reset your password.',
    input: 'email',
    inputPlaceholder: 'Enter your email address',
    heightAuto: false,
    customClass: { title: 'swal-custom-overflow' },
  }).then(({ value: email }) => {
    if (email) {
      pwReset(email).then((result) => {
        ReactSwal.fire({
          title: 'Email Sent!',
          text: 'a link for reset your password has been sent.',
          type: 'success',
          heightAuto: false,
          customClass: { icon: 'swal-custom-overflow' },
        }).then(() => {
          setEmail(email);
        });
      }).catch((err) => {
        console.log('SwalForForgotPW pwReset err: ', err);
      });
    }
  });
};

export const signInValidationSchema = Yup.object({
  email: Yup.string().email(),
  password: Yup.string().min(6, 'password must be at least 6 characters.').required('password is required'),
});

const onSignInSubmit = ({ email, password }, { setSubmitting, resetForm }, push, dispatch, signInWithEmailAndPW) => {
  if (signInWithEmailAndPW) {
    signInWithEmailAndPW(email, password)
      .then(({ user: { email } }) => {
        dispatch({
          type: 'set-user',
          payload: { email },
        });
        setSubmitting(false);
        push('/');
      })
      .catch(({ code, message }) => {
        ReactSwal.fire({
          title: code,
          text: message,
          type: 'error',
          heightAuto: false,
        });
        resetForm();
      });
  }
};

export default ({ history: { push, location } }: IProps) => {
  const { state: { firebase: { signInWithEmailAndPW, pwReset } }, dispatch } = useContext(AppContext);
  const [email, setEmail] = useState(() => {
    if (location && location.state) {
      const { email: locEmail = '' } = location.state;
      return locEmail;
    }
    return '';
  });
  const onClickSignUp = useCallback(() => push('/signup'), []);
  const onClickForgotPW = useCallback(() => SwalForForgotPW(pwReset, setEmail), [pwReset]);

  return (
    <Wrapper>
      <Formik
        initialValues={{
          email: '',
          password: '',
        }}
        validationSchema={signInValidationSchema}
        onSubmit={(values, actions) => onSignInSubmit(values, actions, push, dispatch, signInWithEmailAndPW)}
      >
        {({
          isValid,
          isSubmitting,
        }) => {
          return (
            <SignContainer widthinrem={20}>
              <FormInnerWrapper>
                {signInList.map(({ label, type }) => (
                  <Field
                    key={label}
                    name={label}
                    type={type}
                    width={7}
                    receivedEmail={email}
                    setEmail={setEmail}
                    component={SimpleFormInput}
                  />
                ))}
              </FormInnerWrapper>
              <Button
                type="submit"
                disabled={isSubmitting || !isValid}
                isLoading={isSubmitting}
                text='Sign In'
                height="3.5rem"
              />
              <BottomTextBtnContainer>
                <BottomTextButton
                  onClick={onClickSignUp}
                  value="Sign Up"
                />
                <BottomTextButton
                  right
                  onClick={onClickForgotPW}
                  value="Forgot password?"
                />
              </BottomTextBtnContainer>
            </SignContainer>
          );
        }}
      </Formik>
    </Wrapper>
  );
};
