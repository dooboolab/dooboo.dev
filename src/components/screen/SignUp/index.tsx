import React, { useContext, useCallback } from 'react';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import Swal from 'sweetalert2';
import withReactContent from 'sweetalert2-react-content';

import { AppContext } from '../../../contexts';
import {
  Wrapper,
  SignContainer,
  FormInnerWrapper,
  BottomTextBtnContainer,
  BottomTextButton,
} from './styles';
import { SimpleFormInput } from '../../shared/Input';
import Button from '../../shared/Button';

const ReactSwal = withReactContent(Swal);

export const signUpList = [{
  label: 'email',
  type: 'text',
}, {
  label: 'password1',
  type: 'password',
}, {
  label: 'password2',
  type: 'password',
}];

const signUpValidationSchema = Yup.object({
  email: Yup.string().email(),
  password1: Yup.string().min(6, 'password must be at least 6 characters.').required('password is required'),
  password2: Yup.string().min(6, 'password must be at least 6 characters.').oneOf([Yup.ref('password1'), null], 'passwords must match'),
});

const onSignUpSubmit = ({ email, password1 }, { setSubmitting, resetForm }, history, createUserWithEmailAndPW) => {
  if (createUserWithEmailAndPW) {
    createUserWithEmailAndPW(email, password1)
      .then(({ user: { email } }) => {
        setSubmitting(false);
        ReactSwal.fire({
          title: 'Signing up completed.',
          html: `${email} has been registered in our system.<br />Please sign in now.`,
          type: 'success',
          heightAuto: false,
          customClass: { icon: 'swal-custom-overflow' },
        }).then(() => {
          history.push({
            pathname: '/signin',
            state: { email },
          });
        });
      })
      .catch(({ code, message }) => {
        ReactSwal.fire({
          title: code,
          text: message,
          type: 'error',
          heightAuto: false,
        });
        resetForm();
      });
  }
};

export default ({ history }) => {
  const { state: { firebase: { createUserWithEmailAndPW } } } = useContext(AppContext);
  const onClickSignIn = useCallback(() => history.push('/signin'), []);

  return (
    <Wrapper>
      <Formik
        initialValues={{
          email: '',
          password1: '',
          password2: '',
        }}
        validationSchema={signUpValidationSchema}
        onSubmit={(values, actions) => onSignUpSubmit(values, actions, history, createUserWithEmailAndPW)}
      >
        {({
          isSubmitting,
          isValid,
        }) => (
          <SignContainer widthinrem={20}>
            <FormInnerWrapper>
              {signUpList.map(({ label, type }) => (
                <Field
                  key={label}
                  name={label}
                  type={type}
                  width={7}
                  component={SimpleFormInput}
                />
              ))}
            </FormInnerWrapper>
            <Button
              type="submit"
              disabled={isSubmitting || !isValid}
              isLoading={isSubmitting}
              text="Sign Up"
              height="3.5rem"
            />
            <BottomTextBtnContainer>
              <BottomTextButton
                onClick={onClickSignIn}
                value="Sign In"
              />
            </BottomTextBtnContainer>
          </SignContainer>
        )}
      </Formik>
    </Wrapper>
  );
};
