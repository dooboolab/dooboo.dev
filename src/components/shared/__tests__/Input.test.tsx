import React from 'react';
import { Formik, Form, Field } from 'formik';

import { SimpleFormInput, getLabelCapitalize } from '../Input';
import { signInList, signInValidationSchema } from '../../screen/SignIn';
import { render, fireEvent, wait, waitForElement } from '../../../../test/test-utils';

describe('[SimpleFormInput] render', () => {
  it('renders without crashing', () => {
    const component: any = render(
      <Formik
        initialValues={{ email: '', password: '' }}
        validationSchema={signInValidationSchema}
        onSubmit={() => jest.fn()}
      >
        <Form>
          {signInList.map(({ label, type }) => (
            <Field
              key={label}
              name={label}
              type={type}
              width={7}
              component={SimpleFormInput}
            />
          ))}
        </Form>
      </Formik>
    );
    expect(component.firstChild).toMatchSnapshot();
  });
});

describe('[SimpleFormInput] interaction', () => {
  const component =
    <Formik
      initialValues={{ email: '', password: '' }}
      validationSchema={signInValidationSchema}
      onSubmit={() => jest.fn()}
    >
      <Form>
        {signInList.map(({ label, type }) => (
          <Field
            key={label}
            name={label}
            type={type}
            width={7}
            component={SimpleFormInput}
          />
        ))}
      </Form>
    </Formik>;

  let renderResult: any;

  beforeEach(() => {
    renderResult = render(component);
  });

  it('getLabelCapitalize func works as designed.', async () => {
    const { getByTestId } = renderResult;
    const labelEmail = getByTestId(`label-${signInList[0].label}`);
    expect(labelEmail.textContent).toBe(getLabelCapitalize(signInList[0].label));
  });

  it('ErrorMessage displays as intended: SignIn email and password.', async () => {
    const { findByTestId, getByLabelText } = renderResult;
    const inputEmail = getByLabelText(getLabelCapitalize(signInList[0].label));
    fireEvent.change(inputEmail, { target: { value: 'hello' } });
    fireEvent.blur(inputEmail);
    const emailErrorMsg = await findByTestId(`errors-${signInList[0].label}`);
    expect(emailErrorMsg.textContent).toBe('email must be a valid email');

    const inputPW = getByLabelText(getLabelCapitalize(signInList[1].label));
    fireEvent.change(inputPW, { target: { value: '' } });
    fireEvent.blur(inputPW);
    const pwErrorMsg = await findByTestId(`errors-${signInList[1].label}`);
    expect(pwErrorMsg.textContent).toBe('password is required');

    fireEvent.change(inputPW, { target: { value: 'hello' } });
    await wait(() => expect(pwErrorMsg.textContent).toBe('password must be at least 6 characters.'));
  });
});
