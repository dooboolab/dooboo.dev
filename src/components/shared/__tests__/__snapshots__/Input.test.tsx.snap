// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SimpleFormInput] render renders without crashing 1`] = `undefined`;
